### ISSUE TITLE
Worker Threads: "Module did not self-register"

### ISSUE THREAD

**Issue by mattolson** [schroffl/node-lzo#10]
When using the new experimental worker threads in Node 10.5 and above, if you require lzo in the worker, the following error is thrown: 

```
Error: Module did not self-register.
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:718:18)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:20:18)
    at bindings (/Users/matt.olson/workspace/storefront-renderer/node_modules/bindings/bindings.js:76:44)
    at Object.<anonymous> (/Users/matt.olson/workspace/storefront-renderer/node_modules/lzo/index.js:3:36)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
```

This seems to be related to https://github.com/nodejs/node/issues/21783 and https://github.com/nodejs/node/issues/21481. From the comments, it appears that this is a special case that authors of native addons need to handle. I'm really keen on getting this to work. Let me know if it's already on your radar or if you'd like some help with it. Thanks!

**Comment by schroffl** [comment]
I just tested it with Version `10.12.0` and a fresh copy of this repository:

`abc.js`
```javascript
﻿const { Worker } = require('worker_threads');
const worker = new Worker('./worker.js', { workerData: 'yo' });

worker.on('message', console.log);
worker.on('error', console.error);
worker.on('exit', code => console.log('Worker exit: ', code));
```

`worker.js`
```javascript
const { workerData, parentPort } = require('worker_threads');
const lzo = require('.');

parentPort.postMessage(lzo.compress(workerData));
```

#### Result
```
$ node --version
v10.12.0

$ node --experimental-worker abc.js
Uint8Array [ 19, 121, 111, 17, 0, 0 ]
Worker exit:  0
```

**Comment by schroffl** [comment]
However, I found [this Stackoverflow answer](https://stackoverflow.com/a/49224357/5516238) and in my [`bindings.gyp`](https://github.com/schroffl/node-lzo/blob/master/binding.gyp#L5) the files actually happen to be in the _<q>wrong</q>_ order.
I can try to flip them around and publish a new version if you want to give it a shot :)

**Comment by mattolson** [comment]
I’d be happy to test a new version. Glad to see there is nothing fundamentally wrong.

**Comment by schroffl** [comment]
I will aim to do that later today.

**Comment by schroffl** [comment]
I just published a new version. For some reason it's not yet visible on [the npm website](https://www.npmjs.com/package/lzo), but I can see it [in the registry json](https://registry.npmjs.org/lzo).
Version `0.4.7` should be installed when you update.

**Comment by mattolson** [comment]
I have been able to get a simple test case working with 0.4.7 that uses lzo in a worker thread. However, there's some other complexity in my app that is still leading to the same result. I'll let you know when I track it down. Thanks for the help.

**Comment by mattolson** [comment]
Ok, it turns out that this issue happens when I require the lzo module in both the parent and the worker.

Here is my test code:

parent.js:
```javascript
'use strict';

const util = require('util');
const { Worker } = require('worker_threads');

function sendMessage(worker, operation, input) {
    return new Promise((resolve, reject) => {
        // Process response message
        worker.on('message', response => {
            console.log(`Parent: received ${util.inspect(response)}`);

            if (response.errorMessage) {
                return reject(new Error(response.errorMessage));
            }

            return resolve(response.result);
        });

        // Send message to worker
        const message = {
            operation,
            input,
        };

        console.log(`Parent: sending ${util.inspect(message)}`);
        worker.postMessage(message);
    });
}

//const lzo = require('lzo'); // uncommenting this line causes the issue
const worker = new Worker('./child.js');

const input = 'foobar';
sendMessage(worker, 'compress', input).then(compressed => {
    console.log(`Parent: got compressed result from worker: ${compressed}`);
    sendMessage(worker, 'decompress', compressed).then(decompressed => {
        console.log(`Parent: got decompressed result from worker: ${Buffer.from(decompressed).toString()}`);
        process.exit();
    });
}).catch(err => {
    console.error(`Parent: got error from worker: ${result}`);
    process.exit(1);
});
```

child.js:
```javascript
'use strict';

const lzo = require('lzo');
const util = require('util');
const { parentPort, workerData } = require('worker_threads');

function compress(message) {
    return lzo.compress(message);
}

function decompress(message) {
    return lzo.decompress(message, message.length * 10);
}

function sendResponse(operation, result) {
    const response = {
        operation,
        result,
    };
    console.log(`Child: sending ${util.inspect(response)}`);
    parentPort.postMessage(response);
}

function sendError(operation, err) {
    const response = {
        operation,
        errorMessage: err.message, // can't serialize Errors
    };
    console.log(`Child: sending ${util.inspect(response)}`);
    parentPort.postMessage(response);
}

parentPort.on('message', message => {
    console.log(`Child: received ${util.inspect(message)}`);

    let result;
    switch(message.operation) {
        case 'compress':
            result = compress(message.input);
            sendResponse(message.operation, result);
            break;
        case 'decompress':
            result = decompress(message.input);
            sendResponse(message.operation, result);
            break;
        case 'terminate':
            process.exit(0);
            break;
    }
});
```

Uncommenting the require of lzo in parent.js causes the error, which occurs in the worker thread during the second require.
```
Error: Module did not self-register.
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:718:18)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:20:18)
    at bindings (/Users/matt.olson/workspace/storefront-renderer/node_modules/bindings/bindings.js:76:44)
    at Object.<anonymous> (/Users/matt.olson/workspace/storefront-renderer/node_modules/lzo/index.js:3:36)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
```

**Comment by mattolson** [comment]
I don't know if this applies to you, but this might help? https://github.com/nodejs/node/issues/21783#issuecomment-429637117

**Comment by schroffl** [comment]
That definitely looks like it might help, I will check it out and try to publish a new release next week. Thanks for your research so far :)

**Comment by schroffl** [comment]
This is published in Version [`v0.4.8`](https://www.npmjs.com/package/lzo/v/0.4.8)

**Comment by mattolson** [comment]
Thank you!
